### PR TITLE
LibGfx/ILBMLoader: Add support for 24bit files

### PR DESCRIPTION
**Pull request by warpdesign**
This PR adds support for ILBM `deep` pictures that have 24 bitplanes. Plane order is:

> saved first -----------------------------------------------> saved last
> R0 R1 R2 R3 R4 R5 R6 R7 G0 G1 G2 G3 G4 G5 G6 G7 B0 B1 B2 B3 B4 B5 B6 B7

Screenshot 2023-12-14 at 10 36 58

Note that it seems there is another common 24bit variant that stores the pixel data in a different order: this non standard encoding is not supported.

### CODE REVIEW

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 1]
```diff
@@ -233,7 +250,7 @@ static ErrorOr<ByteBuffer> uncompress_byte_run(ReadonlyBytes data, ILBMLoadingCo
 
     u32 index = 0;
     u32 read_bytes = 0;
-    while (read_bytes < length) {
+    while (read_bytes < length - 1) {
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:253]: That seems quite unrelated to the purpose of this commit. If this is a bugfix, please make it in its own commit.

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 2]
```diff
@@ -400,17 +422,15 @@ static ErrorOr<void> decode_bmhd_chunk(ILBMLoadingContext& context)
     context.chunks_cursor = context.data.slice(sizeof(IFFHeader));
     auto first_chunk = TRY(decode_iff_advance_chunk(context.chunks_cursor));
 
-    if (first_chunk.type != FourCC("BMHD"))
+    if (first_chunk.type != FourCC("BMHD")) {
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:425]: No need for braces here

**Review by LucasChollet — COMMENTED** [review]
I only gave it a quick glance, but I'll review the PR this we

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp (hunk 1 above)
Comment (warpdesign) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:253]: It is a bugfix and unrelated indeed but I stumbled upon it while testing various iff-24 files.

Do you mean an own commit + PR or another commit in this PR?

Comment (AtkinsSJ) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:253]: Separate commit in the same PR is fine.

Comment (warpdesign) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:253]: I created another commit for the fix.

Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:253]: I have a hard-time understanding what is the fundamental issue.
Because we are not reading past the end of the `Bytes`, as we would hit a `VERIFY` when reading, and I would be surprised that the format *always* include an unneeded byte.

Have you tried images with an odd size? All of our test images have an even number of pixels.

Comment (warpdesign) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:253]: Here is an example [file](https://aminet.net/pix/trace/bwing24.lha) (need to uncompress with something supporting lha like `7zip` or `the unarchiver` on mac) that breaks without the patch: it breaks on line 269.

Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:253]: I don't doubt that it allows us to open more images, but I am concerned for regressions.
Because as I said earlier, you're not reading one-byte after then end, a `VERIFY` would have caught you.
So it means that you're stopping one byte earlier every single time, what information contains this last byte? If that's a filler byte, when is it included? Can we explicitly write that we are skipping it/them?

Comment (warpdesign) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:253]: I understand what you mean but am not sure why such patch is needed. From what I could read about the `byterun1` (also called `PackBits` on Mac) compression algorithm:

> There is no way based on the PackBits data to determine the end of the data stream; that is to say, one must already know the size of the compressed or uncompressed data before reading a PackBits data stream to know where it ends.

On `ffmpeg` they end the decoding loop when either the (packed) buffer has been read or the output buffer is filled, see [here](https://ffmpeg.org/doxygen/0.7/libavcodec_2iff_8c-source.html#l00403).

I updated the loop to be:

```cpp
    while (read_bytes < length && index < plane_data_size) {
    // ...
    }
```

And it also fixes the problem and seems more clear to me.

What do you think?

Comment (gmta) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:253]: This conversation doesn't seem to be resolved?

Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:253]: >also called `PackBits` on Mac

That's why it looked familiar 😅, I also implemented it for the TIFF decoder. We should make a common implementation one you port the decoder to `Stream`.
https://github.com/SerenityOS/serenity/blob/b27a62488c7fcf6bc39dfd927388f575192d1565/Userland/Libraries/LibGfx/ImageFormats/TIFFLoader.cpp#L199-L234

That being said, the issue is probably a bug in the encoder, as a single byte doesn't make sense in a PackBits Stream. Maybe a byte is included to keep the alignment on word boundary 🤷‍♂️.
Anyway, your new proposal is much better IMO as it is both more clear and more robust. You just didn't commit it yet 😄

Comment (warpdesign) [reply Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp]: Yes: I wanted your opinion before :) I just pushed the update (I amended the first commit).

File: Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp [hunk 3]
```diff
@@ -189,31 +195,42 @@ static ErrorOr<ByteBuffer> planar_to_chunky(ReadonlyBytes bitplanes, ILBMLoading
...
     u16 height = context.bm_header.height;
     u8 planes = context.bm_header.planes;
-    auto chunky = TRY(ByteBuffer::create_zeroed(static_cast<size_t>(width) * height));
+    size_t buffer_size = static_cast<size_t>(width) * height;
+    // If planes number is 24 we'll store R,G,B components so buffer needs to be 3 times width*height
+    // otherwise we'll store a single 8bit index to the CMAP.
+    if (planes == 24)
+        buffer_size *= 3;
+
+    auto chunky = TRY(ByteBuffer::create_zeroed(static_cast<size_t>(buffer_size)));
```
Comment (LucasChollet) [inline Userland/Libraries/LibGfx/ImageFormats/ILBMLoader.cpp:204]: nit: This `static_cast` is not needed

**Review by LucasChollet — APPROVED** [review]
